commit 5e82830c5d3c5220afcd5e199826e71dff86b423
Author: Sokolov Innokenty <im@qbbr.io>
Date:   Tue Jan 3 12:23:39 2023 +0800

    patch bottommargin

diff --git a/config.def.h b/config.def.h
index 06d0ce3..1e24c54 100644
--- a/config.def.h
+++ b/config.def.h
@@ -11,6 +11,7 @@ static const int systraypinningfailfirst = 1;   /* 1: if pinning fails, display
 static const int showsystray             = 1;   /* 0 means no systray */
 static const int showbar                 = 1;   /* 0 means no bar */
 static const int topbar                  = 1;   /* 0 means bottom bar */
+static const unsigned int bottommargin   = 20;  /* Margin at the bottom for another bar */
 static const char *fonts[]               = { "monospace:size=10" };
 static const char dmenufont[]            = "monospace:size=10";
 static const char col_gray1[]            = "#222222";
diff --git a/dwm.c b/dwm.c
index e8fbdd1..46bc159 100644
--- a/dwm.c
+++ b/dwm.c
@@ -2149,6 +2149,9 @@ updatebarpos(Monitor *m)
 	m->wh = m->mh;
 	if (m->showbar) {
 		m->wh -= bh;
+		m->wh -= bottommargin;
+		if (!m->topbar)
+			m->wy += bottommargin;
 		m->by = m->topbar ? m->wy : m->wy + m->wh;
 		m->wy = m->topbar ? m->wy + bh : m->wy;
 	} else
