commit e05e08308abfcdb9bf4e34a2c2802881374cde2c
Author: Sokolov Innokenty <im@qbbr.io>
Date:   Tue Jan 3 12:26:10 2023 +0800

    patch resetnmaster

diff --git a/dwm.c b/dwm.c
index 46bc159..990eea5 100644
--- a/dwm.c
+++ b/dwm.c
@@ -213,6 +213,7 @@ static void propertynotify(XEvent *e);
 static void quit(const Arg *arg);
 static Monitor *recttomon(int x, int y, int w, int h);
 static void removesystrayicon(Client *i);
+static void resetnmaster(const Arg *arg);
 static void resize(Client *c, int x, int y, int w, int h, int interact);
 static void resizebarwin(Monitor *m);
 static void resizeclient(Client *c, int x, int y, int w, int h);
@@ -1524,6 +1525,13 @@ removesystrayicon(Client *i)
 	free(i);
 }
 
+void
+resetnmaster(const Arg *arg)
+{
+	selmon->nmaster = 1;
+	arrange(selmon);
+}
+
 void
 resize(Client *c, int x, int y, int w, int h, int interact)
 {
